src/m3ninx/doc/types.go

```diff
@@ -125,8 +125,17 @@ type OnIndexSeries interface {
 	// IndexedForBlockStart returns true if the blockStart has been indexed.
 	IndexedForBlockStart(blockStart xtime.UnixNano) bool
 
-	// IndexedRange returns minimum and maximum blockStart values covered by index entry.
+	// ReconciledIndexedForBlockStart returns true if the blockStart has been indexed.
+	// Attempts to retrieve the most recent index entry from the shard if this entry
+	// was never inserted there. If there is an error, returns the value for this entry.
+	// Additionally, returns a boolean to denote if result came from reconciliation.
+	ReconciledIndexedForBlockStart(blockStart xtime.UnixNano) (bool, bool)
```

Comment:
FWIW, this API works, but it'd be much friendlier, I think, if we could just return the `storage.Entry` from the lookup map as an `OnIndexSeries` here. The problem is we'd need to expose methods on this interface for incrementing and decrementing the reader+writer count which ends up feeling equally unfriendly and more error prone.

Reply:
True yeah.

Although things we could do to make that nicer:
```go
type CleanupFn func()

ReconciledOnIndexSeries() (OnIndexSeries, CleanupFn)
```

Then to call it:
```go
onIndexSeries, cleanup := entry.ReconciledOnIndexSeries()
defer cleanup()

// whatever with onIndexSeries
```

Or something which could be more allocation intensive is passing a function/lambda and doing the lookup and the decrement/etc afterwards explicitly (which might need to be dynamically allocated):
```go
WithReconciledOnIndexSeries(callback func (OnIndexSeries))
```

Reply:
I like the cleanup option since that's a convention we follow in a few different places. Going to rework this PR to do that since I think it'll allow us to get rid of specific `Reconciled...` functions.